### ISSUE TITLE
[🐛BUG] 大数据量训练时会卡住

### ISSUE THREAD

**Issue by sxhysj** [RUCAIBox/RecBole#957]
**描述这个 bug**
大数据量训练时会卡住

**如何复现**
在使用大数据量，例如100万条数据的时候，进度条会卡住不动，但是Ctrl+C可以终止程序。

**预期**
希望找到对应的Debug API或者方法。

**屏幕截图**
添加屏幕截图以帮助解释您的问题。（可选）

**链接**
添加能够复现 bug 的代码链接，如 Colab 或者其他在线 Jupyter 平台。（可选）

**实验环境（请补全下列信息）：**
 - 操作系统: [如 Linux, macOS 或 Windows]
- RecBole 版本 0.2.1
 - Python 版本 3.8
- PyTorch 版本 1.9
- cudatoolkit 版本 11.1

**Comment by Sherry-XLL** [comment]
@sxhysj 您好，方便提供一下您的模型、数据集和配置参数吗？

**Comment by sxhysj** [comment]
> @sxhysj 您好，方便提供一下您的模型、数据集和配置参数吗？

###模型：
```
BPR、DeepFM等。模型超参约大，越容易被卡住。数据量约大，越容易被卡住。
```

###数据集：
```
100万行数据，每行50维度，主要类型为token及float。
```

###配置参数：
```
# model config
embedding_size: 32

# dataset config
field_separator: "," #指定数据集field的分隔符
seq_separator: " " #指定数据集中token_seq或者float_seq域里的分隔符
USER_ID_FIELD: Smb #指定用户id域
ITEM_ID_FIELD: rec #指定物品id域
#RATING_FIELD: rating #指定打分rating域
#TIME_FIELD: timestamp #指定时间域
#NEG_PREFIX: neg_ #指定负采样前缀
#指定从什么文件里读什么列，这里就是从ml-1m.inter里面读取user_id, item_id, rating, timestamp这四列
load_col:
  # inter: [user_id, serv_id, rating, timestamp]
  # user: [user_id, claim_count, complain_count, grade_val, mem_create_day]
  inter: [Smb等一系列数据维度，大约50个]


max_user_inter_num: 6

# training settings
epochs: 1000 #训练的最大轮数
train_batch_size: 4096 #训练的batch_size
learner: adam #使用的pytorch内置优化器
learning_rate: 0.001 #学习率
training_neg_sample_num: 1 #负采样数目
eval_step: 1 #每次训练后做evalaution的次数
stopping_step: 10 #控制训练收敛的步骤数，在该步骤数内若选取的评测标准没有什么变化，就可以提前停止了

# evalution settings
eval_setting: RO_RS,full #对数据随机重排，设置按比例划分数据集，且使用全排序
group_by_user: True #是否将一个user的记录划到一个组里，当eval_setting使用RO_RS的时候该项必须是True
split_ratio: [0.8,0.1,0.1] #切分比例
metrics: ["Recall", "MRR","NDCG","Hit","Precision"] #评测标准
topk: [3] #评测标准使用topk，设置成10评测标准就是["Recall@10", "MRR@10", "NDCG@10", "Hit@10", "Precision@10"]
valid_metric: MRR@3 #选取哪个评测标准作为作为提前停止训练的标准
eval_batch_size: 4096 #评测的batch_size
```

**Comment by Sherry-XLL** [comment]
@sxhysj 您好，请问您训练时能完整地跑完一个 epoch 吗？还是直接就卡在第一个 epoch 的进度条不动？如果一个 epoch 都跑不了，可能是设备问题；如果进度条卡住不动，可以检查一下是否超出了设备的显存和内存。

**Comment by sxhysj** [comment]
> @sxhysj 您好，请问您训练时能完整地跑完一个 epoch 吗？还是直接就卡在第一个 epoch 的进度条不动？如果一个 epoch 都跑不了，可能是设备问题；如果进度条卡住不动，可以检查一下是否超出了设备的显存和内存。

可以跑几个epoch，但是会在某个epoch的进度条卡在中间。我检查了GPU显存，大概会用掉1G左右，我的显存还剩9G。内存也够用

**Comment by Sherry-XLL** [comment]
@sxhysj 请问您能不能发送您的数据集到 recbole@outlook.com，我们试着复现一下您的问题，再有针对性地进行解答。

**Comment by sxhysj** [comment]
> @sxhysj 请问您能不能发送您的数据集到 [recbole@outlook.com](mailto:recbole@outlook.com)，我们试着复现一下您的问题，再有针对性地进行解答。

抱歉，数据无法导出。

我在卡住的时候用Ctrl+C中断，添加了Debug信息，发现在Sample.py的150的check_list长度一直在一个固定值"7"循环，也就是说进入了一个死循环，可能bug出在这里。请问有什么办法进一步debug?

**Comment by Sherry-XLL** [comment]
@sxhysj 您可以试着在 `sampler.py` 的 `random_num` 函数的第 90 行之前加一句 `np.random.shuffle(self.random_list)`，看看是否还会出现之前的问题，我们也在下一版本对采样部分进行了更新。

**Comment by sxhysj** [comment]
> @sxhysj 您可以试着在 `sampler.py` 的 `random_num` 函数的第 90 行之前加一句 `np.random.shuffle(self.random_list)`，看看是否还会出现之前的问题，我们也在下一版本对采样部分进行了更新。

通过你说的方法已解决，谢谢！
